9229 - Add bearer token auth mechanism with feature flag

johannes-darms opened the pull request:
**What this PR does / why we need it**:

Implements BearerAccessToken authentication for API requests. Changes are structured similarly to PR #9290 (Session auth), the code itself is based on #9230 (differences are adaptation to the new AuthFilter implementation and delegation of token validation to OIDCAuthProvider). Tokens are implicitly validated by the existing getUserInfo code path from OIDCAuthProvider. This means that for each API request secured with a BearerAccessToken, the ODIC authentication service will be contacted. This could slow down response times, but requires almost no new code. 

**Which issue(s) this PR closes**:

- Closes #9229

**Special notes for your reviewer**:

In short those steps must be performed:
1. Configuration of an OIDC Authentication provider is needed. (cf. http://preview.guides.gdcc.io/en/develop/installation/oidc.html).
2. Create a User Account with OIDC Option
3. Obtain a Bearer Token from  OIDC Authentication provider
4. Perform an API request with BearerAccessToken


Testing of the feature locally with all dependencies in in docker needs some configuration. I did the following: 

**Configuration of docker environment** 
I've added this section to the `docker-compose-dev.yaml` to start keycloack with the other dependencies. Keycloak starts with port 8090 (8080 is already used by dataverse). Furthermore, since dataverse and the web browser must be able to connect to keycloack with the same address, I configured the DNS alias within docker `keycloak.mydomain.com` and modified my `/etc/hosts` accordingly (by appending a line `127.0.0.1 keycloak.mydomain.com`). 

```
  dev_keycloak:
    container_name: "dev_keycloack"
    image: 'quay.io/keycloak/keycloak:19.0'
    hostname: keycloak
    environment:
      - KEYCLOAK_ADMIN=kcadmin
      - KEYCLOAK_ADMIN_PASSWORD=kcpassword
      - KEYCLOAK_LOGLEVEL=DEBUG
      - KC_HOSTNAME_STRICT=false
    networks:
      dataverse:
        aliases:
          - keycloak.mydomain.com #create a DNS alis within the network (add the same alias to your /etc/hosts to get a working OIDC flow)
    command: start-dev --import-realm --http-port=8090  # change port to 8090, so within the network and external the same port is used
    ports:
      - "8090:8090"
    volumes:
      - './conf/keycloak/oidc-realm.json:/opt/keycloak/data/import/oidc-realm.json'
```

Similarly I adapted the file `/conf/keycloak/oidc-realm.json` to use the domain (i.e updated the factory data to `type: oidc | issuer: http://keycloak.mydomain.com:8090/realms/oidc-realm | clientId: oidc-client | clientSecret: ss6gE8mODCDfqesQaSG3gwUwZqZt547E") prior configuring dataverse as described here http://preview.guides.gdcc.io/en/develop/installation/oidc.html.


**Obtain Bearer Token**
In order to obtain an Bearer Token I've enabled `Direct access grants` for the `oidc-client` of the `oidc-realm` in the keycloack admin UI. Then the following snippet can be used to get an Access Token.
```
curl -X POST --location "http://keycloak.mydomain.com:8090/realms/oidc-realm/protocol/openid-connect/token" \
    -H "Content-Type: application/x-www-form-urlencoded" \
    -d "username=kcuser&password=kcpassword&grant_type=password&client_id=oidc-client&client_secret=ss6gE8mODCDfqesQaSG3gwUwZqZt547E"
```
The previously obtained Token can than be used for API request towards dataverse. This requires a dataverse user account created via the configured OIDC Provider.
```
export TOKEN= ""
curl -X POST --location "http://localhost:8080/api/dataverses/root" \
    -H "Accept: application/json" \
    -H "Authorization: Bearer $TOKEN" \
    -H "Content-Type: application/json" \
    -d "{
          \"name\": \"Scientific Research\",
          \"alias\": \"science\",
          \"dataverseContacts\": [
            {
              \"contactEmail\": \"pi@example.edu\"
            },
            {
              \"contactEmail\": \"student@example.edu\"
            }
          ],
          \"affiliation\": \"Scientific Research University\",
          \"description\": \"We do all the science.\",
          \"dataverseType\": \"LABORATORY\"
        }"
```


**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:
No.

**Is there a release notes update needed for this change?**:
Yes.

**Additional documentation**:

GPortas commented on src/test/java/edu/harvard/iq/dataverse/api/auth/doubles/BearereTokenKeyContainerRequestTestFake.java:
```diff
@@ -0,0 +1,23 @@
+
+    private final String apiKey;
+
+    public BearereTokenKeyContainerRequestTestFake(String apiKey) {
```
I think there is a typo here: "Bearere"

johannes-darms replied:
Thanks, fixed it

johannes-darms commented:
@GPortas : hey I've fixed the typo and a markdown error within the config.rst and rebased on development. Is there anything else I can do to ease the review?

pdurbin commented:
@johannes-darms we just moved this PR into the sprint and I was about to review it but I think something has gone wrong with this PR. It's showing 72 changed files, including lots of stuff that I'm sure you didn't touch.

Would you be be able to create a fresh PR from the develop branch? Sorry for the trouble!

pdurbin commented:
@johannes-darms meanwhile I can use this "compare" to see what you changed (only 8 changed files):

https://github.com/IQSS/dataverse/compare/develop...pdurbin:dataverse:9229-bearer-api-auth2

I created that branch from your branch and merged the develop branch into it.

pdurbin commented:
I haven't had a chance to test this yet but some initial feedback:

> I've added this section to the docker-compose-dev.yaml

I don't see this in the PR. Should it be there? Also, a small typo: "alis". Once we have this we can probably remove conf/keycloak/docker-compose.yml (and rewrite the docs a bit).

> 127.0.0.1 keycloak.mydomain.com

Is this simply a preference? What if we just use localhost instead of a custom domain? Will it work? I'd rather not add an extra step of editing /etc/hosts if we don't have to.

> Similarly I adapted the file /conf/keycloak/oidc-realm.json to use the domain (i.e updated the factory data to `type: oidc | issuer: http://keycloak.mydomain.com:8090/realms/oidc-realm | clientId: oidc-client | clientSecret: ss6gE8mODCDfqesQaSG3gwUwZqZt547E")

I think you mean conf/keycloak/oidc-keycloak-auth-provider.json

> In order to obtain an Bearer Token I've enabled Direct access grants for the oidc-client of the oidc-realm in the keycloack admin UI.

It would be nice to not use the GUI for this.

Do you think it's one of these? Can we just flip it to true?

```
cat conf/keycloak/oidc-realm.json | jq '.clients[] | "\(.id) \(.name) directAccessGrantsEnabled=\(.directAccessGrantsEnabled)"'
"aed2e103-ee29-4d5c-a34e-1b8c65b7d537 ${client_account} directAccessGrantsEnabled=false"
"1e821c0e-f6b9-4324-9b23-e82b5431fb72 ${client_account-console} directAccessGrantsEnabled=false"
"dddcc3e0-d742-422b-8b5f-84a292ea9d66 ${client_admin-cli} directAccessGrantsEnabled=true"
"df6f6cd0-a046-492f-84ac-b4fe31909be4 ${client_broker} directAccessGrantsEnabled=false"
"c0af31b9-21aa-4e70-baf3-8d68850c4081 null directAccessGrantsEnabled=false"
"43ffb712-f233-48e2-ae79-d6993bac34a5 ${client_realm-management} directAccessGrantsEnabled=false"
"3747f98f-efbb-49ef-8238-a349bf5ab409 ${client_security-admin-console} directAccessGrantsEnabled=false"
```

poikilotherm commented:
> > 127.0.0.1 keycloak.mydomain.com
> 
> Is this simply a preference? What if we just use localhost instead of a custom domain? Will it work? I'd rather not add an extra step of editing /etc/hosts if we don't have to.

I can answer that! The reason is, once again, containers. When a client reaches out to the backend, the backend will in turn need to reach out to the OIDC provider. 

That can obviously not be configured as "localhost", as it needs to address a different container. So you must configure this to some other hostname (or FQDN) that will be resolvable over the Docker internal DNS server.

At the same time, you need the client to reach out to the OIDC provider as well. Now the name becomes important. For the login to the UI for example, you need to provide an OIDC endpoint URL that is resolvable by the client. Also, an OIDC client will check if the names match - this is an inherent security feature.

The consequence is that you need a hostname that is resolvable on both client (your browser) and within the container. But at the same time, you need to have the container resolve this to the internal OIDC provider container IP and on your client this needs to be equivalent to localhost (because you need to access the exposed ports on your machine).

This is usually only possible if you edit `/etc/hosts`. Sorry.

pdurbin commented:
## before

$ curl -X POST --location "http://keycloak.mydomain.com:8090/realms/oidc-realm/protocol/openid-connect/token"     -H "Content-Type: application/x-www-form-urlencoded"     -d "username=kcuser&password=kcpassword&grant_type=password&client_id=oidc-client&client_secret=ss6gE8mODCDfqesQaSG3gwUwZqZt547E"
{"error":"unauthorized_client","error_description":"Client not allowed for direct access grants"}

## after

$ curl -X POST --location "http://keycloak.mydomain.com:8090/realms/oidc-realm/protocol/openid-connect/token"     -H "Content-Type: application/x-www-form-urlencoded"     -d "username=kcuser&password=kcpassword&grant_type=password&client_id=oidc-client&client_secret=ss6gE8mODCDfqesQaSG3gwUwZqZt547E"
{"access_token":"eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICI0MlBxSWJpaVM4UGEyUFJuRTVMOF9Ub2FDWDhIcnQ0eHlrNHB5ajJZaW1rIn0.eyJleHAiOjE2ODM4MzQwODAsImlhdCI6MTY4MzgzMzc4MCwianRpIjoiNDFjYjlkZTQtMjU5Ny00NzZjLWE2ZjEtYWQ0MDgxZjU2MjYxIiwiaXNzIjoiaHR0cDovL2tleWNsb2FrLm15ZG9tYWluLmNvbTo4MDkwL3JlYWxtcy9vaWRjLXJlYWxtIiwic3ViIjoiYjNkNjMwOGMtNmQyMS00MzQyLTg3ZTMtNDEwNDcyYjVkNzQxIiwidHlwIjoiQmVhcmVyIiwiYXpwIjoib2lkYy1jbGllbnQiLCJzZXNzaW9uX3N0YXRlIjoiYzNiYjk1N2MtMDQyNy00NDgyLTkzMzYtOTlmNDI4MWEwODhkIiwiYWxsb3dlZC1vcmlnaW5zIjpbXSwic2NvcGUiOiJlbWFpbCBwcm9maWxlIiwic2lkIjoiYzNiYjk1N2MtMDQyNy00NDgyLTkzMzYtOTlmNDI4MWEwODhkIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsIm5hbWUiOiJUZXN0IFRlc3QiLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJrY3VzZXIiLCJnaXZlbl9uYW1lIjoiVGVzdCIsImZhbWlseV9uYW1lIjoiVGVzdCIsImVtYWlsIjoidGVzdEB0ZXN0LmNvbSJ9.TL_uI-1zb0SDU5dKKX975Q6r3QLBis1mYN7fcFuuv1ISDmiJ7_n6ScqPtpCPUvAY1JRwYeKwnfuF4KNq8L5E-YnNOvttEGDQ_3UnNstqRr6cRE2XXG1lSk0FFqS6CLmKYnt-90A5674f1dU2k8DhQUgcBnK8jJwxTsialDJjjCcKIexU0Tglr-OH9lCmEgq8-8OpmONDdrTaDHuGmLa2X56CLRhAvTkVPWozvDifJVIrSIAoDz3NS-jXskYC5sIWvrMFKGSx1_WuSEwRtCEe-j2igEIVvDt2E1jlSOQure5_FK_vAezMQenoUzuBwaazo5Hlydx3vejU-14Z6dJX0g","expires_in":300,"refresh_expires_in":1800,"refresh_token":"eyJhbGciOiJIUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICIyZjNjOWQyYi01MmM4LTRiZTAtYWJlMi1jMjg2MTkwN2Q5ZTUifQ.eyJleHAiOjE2ODM4MzU1ODAsImlhdCI6MTY4MzgzMzc4MCwianRpIjoiZDg4MjU3MzUtZDYyYy00MTgyLWI1M2EtZWUwMmFjNjNjZjMzIiwiaXNzIjoiaHR0cDovL2tleWNsb2FrLm15ZG9tYWluLmNvbTo4MDkwL3JlYWxtcy9vaWRjLXJlYWxtIiwiYXVkIjoiaHR0cDovL2tleWNsb2FrLm15ZG9tYWluLmNvbTo4MDkwL3JlYWxtcy9vaWRjLXJlYWxtIiwic3ViIjoiYjNkNjMwOGMtNmQyMS00MzQyLTg3ZTMtNDEwNDcyYjVkNzQxIiwidHlwIjoiUmVmcmVzaCIsImF6cCI6Im9pZGMtY2xpZW50Iiwic2Vzc2lvbl9zdGF0ZSI6ImMzYmI5NTdjLTA0MjctNDQ4Mi05MzM2LTk5ZjQyODFhMDg4ZCIsInNjb3BlIjoiZW1haWwgcHJvZmlsZSIsInNpZCI6ImMzYmI5NTdjLTA0MjctNDQ4Mi05MzM2LTk5ZjQyODFhMDg4ZCJ9.O0KON9u7hQYgo-gT6UDwIaPf7iWwYNmdB4R222PC_gM","token_type":"Bearer","not-before-policy":0,"session_state":"c3bb957c-0427-4482-9336-99f4281a088d","scope":"email profile"}

## what I did

Switched from master to Oidc-realm

![Screen Shot 2023-05-11 at 3 34 57 PM](https://github.com/IQSS/dataverse/assets/21006/f7ed9ed4-d4a9-4fde-952e-cf913a1953d5)

Check the "direct access grants" box (as specified above) and click save

![Screen Shot 2023-05-11 at 3 35 45 PM](https://github.com/IQSS/dataverse/assets/21006/d700afd1-4b9b-4034-8d2e-44d7fdecf105)

pdurbin commented:
Yes, I'm going to try again when I'm back at my desk. I feel like we're close! Thanks for this PR!

I wonder if you can fix this PR. Can you please try merging the latest from "develop" into it. If that doesn't work, a new PR is fine and we can continue the discussion in the new PR. Thanks!

pdurbin commented:
Right, right, when I tried to login as kcuser again, Dataverse is confused and no longer has the association:

![Screen Shot 2023-05-12 at 9 50 37 AM](https://github.com/IQSS/dataverse/assets/21006/f50c1c60-5147-4124-9140-00971069b064)

So I delete kcuser:

$ curl -X DELETE http://localhost:8080/api/admin/authenticatedUsers/kcuser
{"status":"OK","data":{"message":"AuthenticatedUser @kcuser deleted."}}

Then I logged in with kcuser again to create an account in Dataverse.

![Screen Shot 2023-05-12 at 9 55 07 AM](https://github.com/IQSS/dataverse/assets/21006/b3e34495-609a-4971-aa45-e854622e144f)
![Screen Shot 2023-05-12 at 9 55 17 AM](https://github.com/IQSS/dataverse/assets/21006/5c3da248-1a45-48be-81e9-7937d3a8a7ad)

This new account ends up having an id of 3 instead of 2. The account with id 2 was deleted above:

![Screen Shot 2023-05-12 at 9 58 52 AM](https://github.com/IQSS/dataverse/assets/21006/80aa5735-4da8-4e69-bb5b-60364d318eae)
 
Now it works! I can use kcuser's bearer token to get "me" info about that user:

$ export TOKEN=`curl -s -X POST --location "http://keycloak.mydomain.com:8090/realms/oidc-realm/protocol/openid-connect/token"     -H "Content-Type: application/x-www-form-urlencoded"     -d "username=kcuser&password=kcpassword&grant_type=password&client_id=oidc-client&client_secret=ss6gE8mODCDfqesQaSG3gwUwZqZt547E" | jq '.access_token' -r | tr -d "\n"`
$ 
$ echo $TOKEN
eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJNR0RKVkZZaXNCajlTWTY1UUw0aS1MSUl5SWg5S21tY3VWdHNaOE1WZnRRIn0.eyJleHAiOjE2ODM5MDAwNDIsImlhdCI6MTY4Mzg5OTc0MiwianRpIjoiZjIzYWJiODItYjg3OC00ZDliLThhYTEtZDMxNTExN2M4YzM3IiwiaXNzIjoiaHR0cDovL2tleWNsb2FrLm15ZG9tYWluLmNvbTo4MDkwL3JlYWxtcy9vaWRjLXJlYWxtIiwic3ViIjoiNTkxNmFjMDktZDk5NC00MjBiLWIzMjItYWEyYmVmNmY5OTgxIiwidHlwIjoiQmVhcmVyIiwiYXpwIjoib2lkYy1jbGllbnQiLCJzZXNzaW9uX3N0YXRlIjoiMmJiOWZjZDgtOWQ4OC00ODVkLTk2ODgtOGZmNzRjMDMwMDA4IiwiYWxsb3dlZC1vcmlnaW5zIjpbXSwic2NvcGUiOiJlbWFpbCBwcm9maWxlIiwic2lkIjoiMmJiOWZjZDgtOWQ4OC00ODVkLTk2ODgtOGZmNzRjMDMwMDA4IiwiZW1haWxfdmVyaWZpZWQiOnRydWUsIm5hbWUiOiJUZXN0IFRlc3QiLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJrY3VzZXIiLCJnaXZlbl9uYW1lIjoiVGVzdCIsImZhbWlseV9uYW1lIjoiVGVzdCIsImVtYWlsIjoidGVzdEB0ZXN0LmNvbSJ9.Bg8cQw8dKuopNZtJuZJcSFBMZYeWw03iRqQgkYFPT3Acks-vrUidvFCe4a8oTW3JTrywr8i__MaxJJe9Ed2iHO1ecj9ohAyQE2MqzNofw-deuBQAs6_fDfj5QKLrmMErU5YyKzkQ81CGhf4EQLyFFk7USBNwZsEP3g76gHiJWFpEWkFsy5GSBgTUOpeJFfqqEmcyz1CGJYWsGA_jylfjczcHSi1M8msmtdrUTcLKM4OHUzHSy5Peub093Rc9p8ExT53mUn-DjPTUWwmlC0gYASGuNDOMzLM_SOOj985PMATIengECWKvfrGiuNkwCAtFjl_KNSF50CnfbuG1us18ow
$ 
$ curl -s -H "Authorization: Bearer $TOKEN" http://localhost:8080/api/users/:me | jq .
{
  "status": "OK",
  "data": {
    "id": 3,
    "identifier": "@kcuser",
    "displayName": "Kcfirst Kclast",
    "firstName": "Kcfirst",
    "lastName": "Kclast",
    "email": "test@test.com",
    "superuser": false,
    "deactivated": false,
    "persistentUserId": "5916ac09-d994-420b-b322-aa2bef6f9981",
    "createdTime": "2023-05-12T13:55:08Z",
    "lastLoginTime": "2023-05-12T13:55:08Z",
    "lastApiUseTime": "2023-05-12T13:56:04Z",
    "authenticationProviderId": "oidc-keycloak"
  }
}

@johannes-darms great work! Now we just need to get this PR down from 72 changed files to 5 or whatever. A new PR is fine if that's easier.

pdurbin commented on doc/sphinx-guides/source/installation/config.rst:
```diff
@@ -2391,6 +2391,9 @@ please find all known feature flags below. Any of these flags can be activated u
     * - api-session-auth
       - Enables API authentication via session cookie (JSESSIONID). **Caution: Enabling this feature flag exposes the installation to CSRF risks!** We expect this feature flag to be temporary (only used by frontend developers, see `#9063 <https://github.com/IQSS/dataverse/issues/9063>`_) and removed once support for bearer tokens has been implemented (see `#9229 <https://github.com/IQSS/dataverse ...
       - ``Off``
+    * - api-bearer-auth
+      - Enables API authentication via Bearer Token for OIDC User Accounts. **Information: This feature works only for ODIC UserAccounts!**
+      - ``Off``
```
```suggestion
    * - api-bearer-auth
      - Enables API authentication via Bearer Token for OIDC User Accounts. **Information: This feature works only for OIDC UserAccounts!**
      - ``Off``
```

Small typo above ^^.

This is a great start on docs. They can be previewed here: https://dataverse-guide--9532.org.readthedocs.build/en/9532/installation/config.html

As I mentioned elsewhere, I pushed a commit with some additional suggested docs at 96e8e65. Please take a look.

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/api/auth/BearerTokenAuthMechanism.java:
```diff
@@ -0,0 +1,116 @@
+    protected AuthenticationServiceBean authSvc;
+    @Inject
+    protected UserServiceBean userSvc;
+    private static final Logger logger = Logger.getLogger(ApiKeyAuthMechanism.class.getName());
```
```suggestion
    private static final Logger logger = Logger.getLogger(BearerTokenAuthMechanism.class.getCanonicalName());
```

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/api/auth/BearerTokenAuthMechanism.java:
```diff
@@ -0,0 +1,116 @@
+package edu.harvard.iq.dataverse.api.auth;
+
+import com.nimbusds.oauth2.sdk.ParseException;
+import com.nimbusds.oauth2.sdk.token.BearerAccessToken;
```
Online I see it called "bearer token", not "bearer access token" and I was wondering where the former comes from. I supposed it's this BearerAccessToken class.

Should we consider switching to "bearer token" where we can in our own code (throughout this pull request, I mean)? That it, drop "access" from the middle? Shorter. In line with the RFC which has "Bearer Token Usage" in the title: https://datatracker.ietf.org/doc/html/rfc6750

I looked for any discussion about this upstream but didn't find any: https://bitbucket.org/connect2id/oauth-2.0-sdk-with-openid-connect-extensions/src/master/

johannes-darms replied:
Oauth2 and OIDC define and use different kinds of Token, AcessToken is one of them. I assume its come from there  (https://oauth.net/id-tokens-vs-access-tokens/). Nevertheless, I shortened the constants names and strings

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/settings/FeatureFlags.java:
```diff
@@ -30,7 +30,12 @@ public enum FeatureFlags {
+    /**
+     * Enables API authentication via Bearer Token.
+     * @apiNote Raise flag by setting "dataverse.feature.api-bearer-auth"
+     * @since Dataverse 5.14
```
Just a heads up that this may not make it into 5.14 (which we're trying to ship before the community meeting next month). Also, it's a little uncertain if the release after that will be 5.15 or 6.0! 😅

pdurbin commented on src/test/java/edu/harvard/iq/dataverse/api/auth/doubles/BearerTokenKeyContainerRequestTestFake.java:
```diff
@@ -0,0 +1,20 @@
+package edu.harvard.iq.dataverse.api.auth.doubles;
```
I don't understand what this "doubles" package. It seems to have been added here: https://github.com/IQSS/dataverse/pull/9303/commits/c766afdc2e6302a4f81fe146cf8ee2b5f25b6dd4

I'm also not sure why the two test classes you're adding go into different packages:

![Screen Shot 2023-05-12 at 2 34 56 PM](https://github.com/IQSS/dataverse/assets/21006/feb77a8f-da3d-4c99-8b8d-d73d710b2ce3)

Anyway, don't worry about. I'm asking in Slack: https://iqss.slack.com/archives/C010LA04BCG/p1683916961295199

johannes-darms replied:
I used  https://github.com/IQSS/dataverse/pull/9290  as a skeleton for the MR, the structure is defined there. The double is a ContainerRequest which gets passed to the SUT. I believe that @GPortas added a subpackage to distinguish between the doubles and actual test code.

GPortas replied:
@pdurbin 

_"Test Double is a generic term for any case where you replace a production object for testing purposes"_. (Find out more on: https://martinfowler.com/bliki/TestDouble.html). 

The sub-package "doubles" contains all auth related test doubles, to organize the code. Keeping the top package just for the tests.

There are different types of test doubles, the well-known mocks and stubs are considered doubles. Fake is another type of test double. From the source shared above:

_"Fake: objects actually have working implementations, but usually take some shortcut which makes them not suitable for production"_

pdurbin reviewed (commented):
As I explained in comments in the PR, I tested this and it seems to work as advertised!

I've talking with @johannes-darms about how there's weird git stuff happening in this PR. Commits, from master, etc. He's going to make a fresh PR when he gets a chance.

Meanwhile, @johannes-darms please check out the comments I'm leaving in this review. I'm especially interested in what you think about the config and docs suggestions I made in 96e8e65 (@poikilotherm and @GPortas we should figure out what we want to do with the two docker compose files in play and associated docs).

@johannes-darms I also started a thread in Slack (that I know you can't see) at https://iqss.slack.com/archives/C04EE66HW2Y/p1683900239188549 about what the next steps should be. Does this repo need more PRs? Where do we start working on PKCE flow? Do you have any frontend code that can make immediate use of this PR? Our plan for our new React frontend is to use this bearer token stuff and PKCE but we don't yet have an issue about this at https://github.com/IQSS/dataverse-frontend/issues (because we're focused on an MVP to show at the community meeting).

poikilotherm commented:
> I've talking with @johannes-darms about how there's weird git stuff happening in this PR. Commits, from master, etc. He's going to make a fresh PR when he gets a chance.

It is not the most favorable move, but certainly something went wrong here. So maybe creating a fresh branch and force pushing them to the PR branch would avoid opening a new PR and keep the discussion less scattered.

johannes-darms commented:
I've created a new and clean PR #9532. I've cherry-picked the changed from me and Phil. I hope that helps

>Meanwhile, @johannes-darms please check out the comments I'm leaving in this review. I'm especially interested in what you think about the config and docs suggestions I made in https://github.com/IQSS/dataverse/commit/96e8e657598fb3566179573e7a45606238365bde (@poikilotherm and @GPortas we should figure out what we want to do with the two docker compose files in play and associated docs

I've integrated those changes. Yet, I'm not a fan of having documentation or code that requires changes of the host system within a project. A developer without  sufficient rights to alter the configuration (/etc/hosts) cannot work with the project. Hence, I'd rather add a reverse-proxy (like traefik) into the mix to avoid that dependency on the host system or just document whats needed to set something up. Adding a  reverse proxy makes only sense if it is also used for other purposes. Besides this brings the mainline close to `dataverse-docker` whats already done by...

pdurbin commented:
@johannes-darms thanks for the new PR! I'm closing this one in favor of it:

- #9591
